Add blog post for spatialsample 0.2.0 (#585)
tidyverse/tidyverse.org: content/blog/spatialsample-0-2-0/index.Rmd

---
output: hugodown::hugo_document

slug: spatialsample-0-2-0
title: spatialsample 0.2.0
date: 2022-06-15
author: Mike Mahoney
description: >
    spatialsample 0.2.0 is now on CRAN! This release provides a bunch of new features, including new spatial resampling methods, visualization helpers, and spatial buffering.

photo:
  url: https://unsplash.com/photos/1-29wyvvLJA
  author: Andrew Neel

# one of: "deep-dive", "learn", "package", "programming", "roundup", or "other"
categories: [package] 
tags: [spatialsample, rsample, tidymodels]
---

<!--
TODO:
* [x] Look over / edit the post's title in the yaml
* [x] Edit (or delete) the description; note this appears in the Twitter card
* [x] Pick category and tags (see existing with `hugodown::tidy_show_meta()`)
* [x] Find photo & update yaml metadata
* [x] Create `thumbnail-sq.jpg`; height and width should be equal
* [x] Create `thumbnail-wd.jpg`; width should be >5x height
* [x] `hugodown::use_tidy_thumbnails()`
* [x] Add intro sentence, e.g. the standard tagline for the package
* [x] `usethis::use_tidy_thanks()`
-->

```{r setup, include = FALSE}
library(ggplot2)
theme_set(theme_minimal())
```

> We're positively electrified to announce the release of [spatialsample](https://spatialsample.tidymodels.org/) 0.2.0. spatialsample is a package for spatial resampling, extending the rsample framework for resampling to help create spatial extrapolation between your analysis and assessment data sets.

Reviewer: ```suggestion
We're positively electrified to announce the release of [spatialsample](https://spatialsample.tidymodels.org/) 0.2.0. spatialsample is a package for spatial resampling, extending the [rsample](https://rsample.tidymodels.org/) framework to help create spatial extrapolation between your analysis and assessment data sets.
```

You can install it from CRAN with:

```{r, eval = FALSE}
install.packages("spatialsample")
```

> This blog post will describe the highlights of what's new. You can see a full list of changes in the [release notes](https://github.com/tidymodels/spatialsample/blob/main/NEWS.md).

Reviewer: it'd be nice to update this link to the `0.2.0` section in the NEWS of the pkgdown site, once it's published. then it goes to the right place permanently.

Author: Thanks!

## New Features

This version of spatialsample includes a new data set, made up of 682 hexagons containing data about tree canopy cover change in Boston, Massachusetts:

```{r}
#| fig.alt = "A map showing the spatial arrangement of hexagons making up the boston_canopy data set."
library(spatialsample)
library(ggplot2)

ggplot(boston_canopy) + 
  geom_sf()
```

> This data is stored as an sf object, and as such contains information about the proper coordinate reference system and units of measurement associated with the data. 

Reviewer: Do we have a tidyverse/tidymodels convention about using data as singular or plural?

Reviewer 2: We have a, um, diversity of opinion on this! 😆 I think singular is fine.

> And that brings us to the first new feature in this release of spatialsample: `spatial_clustering_cv()` now supports sf objects, and will calculate distances in a way that respects coordinate reference systems (including using the s2 geometry library for geographic coordinate reference systems):

Reviewer: ```suggestion
This brings us to the first new feature in this release of spatialsample: `spatial_clustering_cv()` now supports sf objects, and will calculate distances in a way that respects coordinate reference systems (including using the s2 geometry library for geographic coordinate reference systems):
```

```{r}
set.seed(123)
kmeans_clustering <- spatial_clustering_cv(boston_canopy, v = 5)
kmeans_clustering
```

This release also provides `autoplot()` methods to visualize resamples via ggplot2, making it easy to see how exactly your data is being divided. Just call `autoplot()` on the outputs from any spatial clustering function:

```{r}
#| fig.alt = "A map showing the boston_canopy data set broken into five folds through spatial_clustering_cv. The five folds are visibly different sizes, and are grouped by spatial proximity."
autoplot(kmeans_clustering) + labs(title = "kmeans()")
```

> In addition to supporting more types of data, `spatial_clustering_cv()` has also been extended to support more types of clustering. Use the `cluster_function` argument to use hierarchical clustering via `hclust()` instead of the default `kmeans()`-based clusters:

Reviewer: ```suggestion
In addition to supporting more types of data, `spatial_clustering_cv()` has also been extended to support more types of clustering. Set the `cluster_function` argument to `"hclust"` for hierarchical clustering via `hclust()` instead of the default `kmeans()`-based clusters:
```

set.seed(123)
spatial_clustering_cv(
  boston_canopy, 
  v = 5, 
  cluster_function = "hclust"
) |> 
  autoplot() + 
  labs(title = "hclust()")
```

This argument can also accept functions, letting you plug in clustering methodologies from other packages or that you've written yourself:

```{r}
#| fig.alt = "A map showing the outputs of spatial_clustering_cv when using a custom clustering function. The custom clustering function assigned folds systematically, moving sequentially through rows in the data frame, and as such the output does not look very clustered. However, the functions in spatialsample performed exactly the same with the custom clustering function as they did with the built-in options."
set.seed(123)

custom_clusters <- function(dists, v, ...) {
  rep(letters[1:v], length.out = nrow(boston_canopy))
}

spatial_clustering_cv(
  boston_canopy, 
  v = 5, 
  cluster_function = custom_clusters
) |> 
  autoplot() + 
  labs(title = "custom_clusters()")
```

> In addition to the clustering extensions, this version of spatialsample introduces a few functions for other popular spatial resampling methods. For instance, the new function `spatial_block_cv()` helps you perform [block cross-validation](https://doi.org/10.1111/ecog.02881), splitting your data into folds based on a grid of regular polygons. You can assign these polygons to folds at random:

Reviewer: ```suggestion
In addition to the clustering extensions, this version of spatialsample introduces new functions for other popular spatial resampling methods. For instance, `spatial_block_cv()` helps you perform [block cross-validation](https://doi.org/10.1111/ecog.02881), splitting your data into folds based on a grid of regular polygons. You can assign these polygons to folds at random:
```

```{r, animation.hook='gifski'}
#| fig.alt = "An animation showing maps of each individual fold produced using spatial_buffer_vfold_cv. Now it is evident that any data adjacent to the assessment data has been added to a 'buffer' zone, and is part of neither the analysis or the assessment set."

blocks$splits |> 
  purrr::walk(function(x) print(autoplot(x)))
```

In addition to exclusion buffers, spatialsample now lets you add inclusion radii to any spatial resampling. This will add any points within a certain distance of the original assessment set to the assessment set, letting you create clumped "discs" of data to assess your models against:

```{r, animation.hook='gifski'}
#| fig.alt = "Another animation showing maps of each individual fold produced using spatial_buffer_vfold_cv. When using the argument radius, points adjacent to the assessment set are themselves added to the assessment set. The buffer is then applied to each data point in the enlarged assessment set."

set.seed(123)
blocks <- spatial_buffer_vfold_cv(
  boston_canopy, 
  v = 20,
  buffer = 100,
  radius = 100
)

blocks$splits |> 
  purrr::walk(function(x) print(autoplot(x)))
```

## ...and more!

> This is just scratching the surface of the new features and improvements in this release of spatialsample. You can see a full list of changes in the the [release notes](https://github.com/tidymodels/spatialsample/blob/main/NEWS.md).

Reviewer: same comment about the link 👀
